lib/Core/Searcher.cpp

```diff
@@ -781,3 +784,145 @@ void InterleavedSearcher::printName(llvm::raw_ostream &os) {
     searcher->printName(os);
   os << "</InterleavedSearcher>\n";
 }
+
+//
+
+ExecutionState &DiscreteTimeFairSearcher::selectState() {
+  if (searchers.find(std::nullopt) != searchers.end()) {
+    return searchers.at(std::nullopt)->selectState();
+  }
+
+  if (!currentQuant || searchers.find(currentFunction) == searchers.end()) {
+    unsigned size = functions.size();
+    currentFunction = functions[theRNG.getInt32() % size];
+    currentQuant = selectQuant;
+  } else {
+    --currentQuant;
+  }
+
+  auto &function = currentFunction;
+  assert(searchers.find(function) != searchers.end() &&
+         searchers.at(function) && !searchers.at(function)->empty());
+  return searchers.at(function)->selectState();
+}
+
+void DiscreteTimeFairSearcher::update(
+    ExecutionState *current, const std::vector<ExecutionState *> &addedStates,
+    const std::vector<ExecutionState *> &removedStates) {
+  if (current) {
+    if (current && (std::find(removedStates.begin(), removedStates.end(),
+                              current) == removedStates.end())) {
+      localStates.insert(current);
+    }
+    for (const auto state : removedStates) {
+      if (std::find(addedStates.begin(), addedStates.end(), state) ==
+          addedStates.end()) {
+        localStates.insert(state);
+      }
+    }
+    update(localStates);
+    localStates.clear();
+  }
+
+  if (current) {
+    auto function = current->multiplexKF
+                        ? std::optional<KFunction *>(current->multiplexKF)
+                        : std::nullopt;
+    localFunction.insert(function);
+    currFunction.insert(function);
+    statesToFunction[current] = function;
+  }
+
+  for (const auto state : addedStates) {
+    auto function = state->multiplexKF
+                        ? std::optional<KFunction *>(state->multiplexKF)
+                        : std::nullopt;
+    localFunction.insert(function);
+    addedKStates[function].push_back(state);
+    statesToFunction[state] = function;
+  }
+
+  for (const auto state : removedStates) {
+    auto function = state->multiplexKF
+                        ? std::optional<KFunction *>(state->multiplexKF)
+                        : std::nullopt;
+    localFunction.insert(function);
+    removedKStates[function].push_back(state);
+    statesToFunction.erase(state);
+  }
+
+  for (auto function : localFunction) {
+
+    ExecutionState *currTState =
+        currFunction.count(function) != 0 ? current : nullptr;
+
+    if (!isThereKFunction(function)) {
+      addKFunction(function);
+    }
+
+    searchers.at(function)->update(currTState, addedKStates[function],
```

Comment:
`at` is a checked version of `[]`. May has a little affect on performance

_(In `update` below you use `[]` :))_